cwl/src/main/scala/cwl/InputParameter.scala

```diff
@@ -84,4 +87,59 @@ object InputParameter {
       }
     }
   }
+
+  /**
+    * Yet another value mapper. This one is needed because in CWL we might need to "augment" inputs which we can only do
+    * once they have been linked to a WomValue. This input value mapper encapsulates logic to be applied once that is
+    * done. For now, if the inputParameter has an input binding with loadContents = true, load the content of the file.
+    *
+    * This is based on the spec in http://www.commonwl.org/v1.0/CommandLineTool.html#Input_binding
+    *
+    * NOTE: There may be _many_ cases not implemented here that need to be fixed.
+    */
+  def inputValueMapper(inputParameter: InputParameter,
+                       inputType: MyriadInputType,
+                       expressionLib: ExpressionLib): InputValueMapper = {
+    ioFunctionSet: IoFunctionSet => {
+
+      def populateFiles(womValue: WomValue): ErrorOr[WomValue] = {
+        womValue match {
+          case womMaybePopulatedFile: WomMaybePopulatedFile =>
+            val parameterContext = ParameterContext(self = womMaybePopulatedFile)
+            val secondaryFilesFromInputParameter = inputParameter.secondaryFiles
+            val secondaryFilesFromType = inputType.fold(MyriadInputTypeToSecondaryFiles)
+            val secondaryFiles = secondaryFilesFromInputParameter orElse secondaryFilesFromType
+            for {
+              loaded <- maybeLoadContents(womMaybePopulatedFile, ioFunctionSet, inputParameter.loadContents)
+              secondaries <- FileParameter.secondaryFiles(
+                loaded,
+                WomSingleFileType,
+                secondaryFiles,
+                parameterContext,
+                expressionLib
+              )
+              updated = loaded.copy(secondaryFiles = secondaries)
+            } yield updated
+
+          case WomArray(_, values) => values.toList.traverse(populateFiles).map(WomArray(_))
+          case womValue: WomValue =>
+            womValue.valid
+        }
+      }
+
+      womValue => populateFiles(womValue).toTry(s"loading $womValue for ${inputParameter.id}").get
+    }
+  }
+
+  private def maybeLoadContents(womMaybePopulatedFile: WomMaybePopulatedFile,
+                                ioFunctionSet: IoFunctionSet,
+                                loadContents: Boolean): ErrorOr[WomMaybePopulatedFile] = {
+    if (loadContents) {
+      FileParameter.load64KiB(womMaybePopulatedFile.value, ioFunctionSet) map { contents =>
```

Comment:
I wonder if we should check if `womMaybePopulatedFile` already has its content loaded for some reason, and if that's the case skip re-loading it here ?

Reply:
At some point, absolutely... I'll file a new ticket for our missing functionality on "File literals" where the content is already provided. In "File literals" there is no file path, we're supposed to make up a temp path. That special handling could also overlap here with skipping re-loading.